File: leetcode/no0088_Merge_Sorted_Array.js
/**
 * @param {number[]} nums1
 * @param {number} m
 * @param {number[]} nums2
 * @param {number} n
 * @return {void} Do not return anything, modify nums1 in-place instead.
 */
var merge = function (nums1, m, nums2, n) {
    const DBG = true;
    //arr.splice(index, 0, item);
    if (DBG) console.log("INPUT: ", nums1);

    let lastIndex = m;
    for (let i = 0; i < nums2.length; i++) {
        let numA = nums2[i];
        if (DBG) console.log("i: ", i, "/", nums2.length);
        let swapped = false;
        for (let j = 0; j < nums1.length; j++) {
            let numB = nums1[j];
            if (DBG) console.log("check", numA, "<", numB);
            if (numA < numB) {
                swapped = true;
                nums1.pop();
                nums1.splice(j, 0, numA);
                lastIndex += 1
                console.log("lastIndex:", lastIndex)
                if (DBG) console.log(`[${j}]: ${numA} --> `, nums1);
                break;
            }
        }
        if (!swapped) {
            if (DBG) console.log("before: ", nums1, `<-- ${numA}, lastIndex:${lastIndex}`);
            nums1[lastIndex] = numA
            lastIndex += 1
            console.log("lastIndex:", lastIndex)
            if (DBG) console.log(" after: ", nums1, `<-- ${numA}`);
        }
        if (DBG) console.log("-----");
    }

    if (DBG) console.log("OUTPUT: ", nums1);
};

let nums1 = [1, 2, 3, 0, 0, 0], m = 3, nums2 = [2, 5, 6], n = 3;
//let nums1 = [1, 2, 3, 0, 0, 0], m = 3, nums2 = [4, 5, 6], n = 3;
//let nums1 = [1], m = 1, nums2 = [], n = 0

merge(nums1, m, nums2, n);
// ans: Output: [1,2,2,3,5,6]

File: leetcode/no0088_Merge_Sorted_Array.java
import java.util.ArrayList;
import java.util.Arrays;

class Solution {

    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int indexB = 0;
        ArrayList<Integer> result = new ArrayList<Integer>();
        for (int i = 0; i < m; i++) {

            int numA = nums1[i];
            boolean hit = false;
            for (int j = indexB; j < nums2.length; j++) {
                int numB = nums2[j];
                if (numA > numB) {
                    hit = true;
                    result.add(numB);
                    indexB += 1;
                    i -= 1;
                    break;
                }
            }

            if (!hit) {
                result.add(numA);
            }
        }

        if (indexB < n) {
            for (int j = indexB; j < nums2.length; j++) {
                int numB = nums2[j];
                result.add(numB);
            }
        }

        for (int i = 0; i < result.size(); i++) {
            nums1[i] = result.get(i);
        }
    }

}

class Main {
    public static void main(String[] args) {
        int[] nums1 = { 1, 2, 3, 0, 0, 0 };
        int[] nums2 = { 2, 5, 6 };

        System.out.println("nums1:" + Arrays.toString(nums1));
        System.out.println("nums2:" + Arrays.toString(nums2));
        Solution s = new Solution();
        s.merge(nums1, 3, nums2, 3);
        System.out.println("answer:" + Arrays.toString(nums1));
    }
}
